src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs

```diff
@@ -1032,29 +1032,98 @@ internal static double Map(double rangeMin, double rangeMax, double inputValue)
     }
     internal class RangeExpressionUtils
     {
-        // For to include start and end. 
-        internal static StackValue[] GenerateRangeByStepNumber(decimal start, decimal end, int stepnum, bool isIntRange)
+      
+        //overloads to avoid returning object[] and boxing/unboxing.
+        internal static long[] GenerateRangeByStepNumberInt(decimal start, decimal end, long stepnum)
         {
+            //TODO watch stepnum - could this cause type issues with floating points?
             decimal stepsize = (stepnum == 1) ? 0 : (end - start) / (stepnum - 1);
-            isIntRange = isIntRange && (Math.Truncate(stepsize) == stepsize);
 
-            StackValue[] range = new StackValue[stepnum > 1 ? stepnum : 1];
-            range[0] = isIntRange ? StackValue.BuildInt((int)start) : StackValue.BuildDouble((double)start);
+            long[] range = new long[stepnum > 1 ? stepnum : 1];
+            range[0] = (long)start;
 
             decimal cur = start;
             for (int i = 1; i < stepnum - 1; ++i)
             {
                 cur += stepsize;
-                range[i] = isIntRange ? StackValue.BuildInt((int)cur) : StackValue.BuildDouble((double)cur);
+                range[i] = (long)cur;
             }
 
             if (stepnum > 1)
             {
-                range[(int)stepnum - 1] = isIntRange ? StackValue.BuildInt((int)end) : StackValue.BuildDouble((double)end);
+                range[stepnum - 1] = (long)end;
             }
 
             return range;
         }
+        internal static double[] GenerateRangeByStepNumberDouble(decimal start, decimal end, long stepnum)
+        {
+            //TODO watch stepnum - could this cause type issues with floating points?
+            decimal stepsize = (stepnum == 1) ? 0 : (end - start) / (stepnum - 1);
+
+            double[] range = new double[stepnum > 1 ? stepnum : 1];
+            range[0] = (double)start;
+
+            decimal cur = start;
+            for (int i = 1; i < stepnum - 1; ++i)
+            {
+                cur += stepsize;
+                range[i] = (double)cur;
+            }
+
+            if (stepnum > 1)
+            {
+                range[stepnum - 1] = (double)end;
+            }
+
+            return range;
+        }
+
+
+        // For to include start and end. 
+        internal static StackValue[] GenerateRangeByStepNumber(decimal start, decimal end, int stepnum, bool isIntRange)
+        {
+            if (isIntRange)
+            {
+                return GenerateRangeByStepNumberInt(start, end, stepnum).Select(x => StackValue.BuildInt(x)).ToArray();
+            }
+            else
+            {
+                return GenerateRangeByStepNumberDouble(start, end, stepnum).Select(x => StackValue.BuildDouble(x)).ToArray();
+            }
+        }
+
+      
+        internal static long[] RangeExpressionInt(long start, long end, double step,int op,bool hasStep, bool hasAmountOp )
```

Comment:
If the step is `double`, e.g. `1..5..1.5`, won't the generated array be a double array?

Reply:
not in all cases - there is a specific existing case for the form start..#amount..step - where step can be a double, but if the value can be represented by an integer than the range is an int array.

Reply:
yes, for now I will implement a fallback boxing case.